File: pluk/src/mcp/namespace.ts
import type { z } from "zod";
import type {
  CallToolResult,
  GetPromptResult,
  McpServer,
  ReadResourceCallback,
  ResourceMetadata,
  ToolAnnotations,
} from "@modelcontextprotocol/server";

// A group exposes several integrations through one MCP server. Their tool/prompt/
// resource names collide (two SQL DBs both register "query"), so in group mode we
// register each member through a namespaced host that prefixes every name with a
// per-member slug. Single-integration endpoints register on the bare McpServer
// and are unaffected.

type Args<S extends z.ZodRawShape> = z.infer<z.ZodObject<S>>;

/**
 * The subset of the MCP server surface an adapter uses to register its own tools,
 * prompts and resources. Positional by design: the SDK's own registration API is
 * config-object shaped (`registerTool(name, { description, inputSchema }, cb)`),
 * and `toolHost()` below is the single place that translates — so an SDK signature
 * change costs one function here, not an edit in every adapter.
 */
export interface ToolHost {
  tool<S extends z.ZodRawShape>(
    name: string,
    description: string,
    schema: S,
    annotations: ToolAnnotations,
    cb: (args: Args<S>, extra?: unknown) => CallToolResult | Promise<CallToolResult>,
  ): unknown;
  tool<S extends z.ZodRawShape>(
    name: string,
    description: string,
    schema: S,
    cb: (args: Args<S>, extra?: unknown) => CallToolResult | Promise<CallToolResult>,
  ): unknown;
  tool(
    name: string,
    description: string,
    annotations: ToolAnnotations,
    cb: (extra?: unknown) => CallToolResult | Promise<CallToolResult>,
  ): unknown;
  tool(
    name: string,
    description: string,
    cb: (extra?: unknown) => CallToolResult | Promise<CallToolResult>,
  ): unknown;
  prompt<S extends z.ZodRawShape>(
    name: string,
    description: string,
    schema: S,
    cb: (args: Args<S>) => GetPromptResult | Promise<GetPromptResult>,
  ): unknown;
  prompt(
    name: string,
    description: string,
    cb: () => GetPromptResult | Promise<GetPromptResult>,
  ): unknown;
  resource(name: string, uri: string, config: ResourceMetadata, cb: ReadResourceCallback): unknown;
  resource(name: string, uri: string, cb: ReadResourceCallback): unknown;
}

// A `tool()` call passes an optional zod shape and optional annotations before the
// callback, so the middle arguments are told apart by shape: annotations carry only
// the spec's hint keys, a zod shape carries anything else.
const ANNOTATION_KEYS = new Set(["title", "readOnlyHint", "destructiveHint", "idempotentHint", "openWorldHint"]);

function isAnnotations(value: unknown): boolean {
  if (typeof value !== "object" || value === null) return false;
  const keys = Object.keys(value);
  return keys.length > 0 && keys.every((k) => ANNOTATION_KEYS.has(k));
}

/** Adapt a real McpServer to the positional `ToolHost` shape adapters register against. */
export function toolHost(server: McpServer): ToolHost {
  const reg = server as unknown as {
    registerTool: (name: string, config: unknown, cb: unknown) => unknown;
    registerPrompt: (name: string, config: unknown, cb: unknown) => unknown;
    registerResource: (name: string, uri: string, config: unknown, cb: unknown) => unknown;
  };
  return {
    tool: ((name: string, description: string, ...rest: unknown[]) => {
      const cb = rest[rest.length - 1];
      const middle = rest.slice(0, -1);
      const annotations = middle.find(isAnnotations);
      const inputSchema = middle.find((m) => !isAnnotations(m));
      return reg.registerTool(name, { description, inputSchema, annotations }, cb);
    }) as ToolHost["tool"],
    prompt: ((name: string, description: string, schemaOrCb: unknown, maybeCb?: unknown) =>
      typeof schemaOrCb === "function"
        ? reg.registerPrompt(name, { description }, schemaOrCb)
        : reg.registerPrompt(name, { description, argsSchema: schemaOrCb }, maybeCb)) as ToolHost["prompt"],
    resource: ((name: string, uri: string, configOrCb: unknown, maybeCb?: unknown) =>
      typeof configOrCb === "function"
        ? reg.registerResource(name, uri, {}, configOrCb)
        : reg.registerResource(name, uri, configOrCb, maybeCb)) as ToolHost["resource"],
  };
}

/** Slugify a member name into a tool-name-safe prefix segment. */
export function slug(name: string): string {
  return name.toLowerCase().replace(/[^a-z0-9]+/g, "_").replace(/^_+|_+$/g, "") || "member";
}

/** Prefix a resource URI so two members' URIs (e.g. `schema://full`) stay unique. */
function namespaceUri(ns: string, uri: string): string {
  const sep = uri.indexOf("://");
  if (sep === -1) return `${ns}+${uri}`;
  return `${uri.slice(0, sep)}://${ns}/${uri.slice(sep + 3)}`;
}

/**
 * Wrap a host so tool/prompt/resource registrations are prefixed with `ns`.
 * Names become `${ns}__${name}`; resource URIs are namespaced too.
 */
export function namespacedHost(host: ToolHost, ns: string): ToolHost {
  const prefix = (name: string) => `${ns}__${name}`;
  return {
    tool: ((name: string, ...rest: unknown[]) =>
      (host.tool as (...a: unknown[]) => unknown)(prefix(name), ...rest)) as ToolHost["tool"],
    prompt: ((name: string, ...rest: unknown[]) =>
      (host.prompt as (...a: unknown[]) => unknown)(prefix(name), ...rest)) as ToolHost["prompt"],
    resource: ((name: string, uri: string, ...rest: unknown[]) =>
      (host.resource as (...a: unknown[]) => unknown)(prefix(name), namespaceUri(ns, uri), ...rest)) as ToolHost["resource"],
  };
}

File: crates/pluk-server/src/mcp/namespace.rs
//! Group namespacing: slug prefixes for tool/prompt/resource names.
//!
//! A group exposes several integrations through one MCP server. Their
//! tool/prompt/resource names collide (two SQL DBs both register `query`), so
//! in group mode each member registers through a [`namespaced`] host that
//! prefixes every name with a per-member slug. Single-integration endpoints
//! register on the bare builder and are unaffected.
//!
//! Ported from `pluk/src/mcp/namespace.ts`.

use pluk_adapters::ToolHost;

/// Slugify a member name into a tool-name-safe prefix segment.
pub fn slug(name: &str) -> String {
    let lowered = name.to_lowercase();
    let mut out = String::with_capacity(lowered.len());
    let mut last_was_sep = false;
    for ch in lowered.chars() {
        if ch.is_ascii_lowercase() || ch.is_ascii_digit() {
            out.push(ch);
            last_was_sep = false;
        } else if !last_was_sep && !out.is_empty() {
            out.push('_');
            last_was_sep = true;
        }
    }
    while out.ends_with('_') {
        out.pop();
    }
    if out.is_empty() { "member".into() } else { out }
}

/// Prefix a resource URI so two members' URIs (e.g. `schema://full`) stay unique.
pub fn namespace_uri(ns: &str, uri: &str) -> String {
    match uri.find("://") {
        Some(sep) => format!("{}://{}/{}", &uri[..sep], ns, &uri[sep + 3..]),
        None => format!("{ns}+{uri}"),
    }
}

/// Wrap a host so tool/prompt/resource registrations are prefixed with `ns`.
/// Names become `<ns>__<name>`; resource URIs are namespaced too.
pub struct NamespacedHost<'a> {
    inner: &'a mut dyn ToolHost,
    ns: String,
}

impl<'a> NamespacedHost<'a> {
    pub fn new(inner: &'a mut dyn ToolHost, ns: impl Into<String>) -> Self {
        NamespacedHost { inner, ns: ns.into() }
    }

    fn prefix(&self, name: &str) -> String {
        format!("{}__{}", self.ns, name)
    }
}

impl ToolHost for NamespacedHost<'_> {
    fn register_tool(
        &mut self,
        registration: pluk_adapters::ToolRegistration,
        handler: pluk_adapters::ToolHandler,
    ) {
        let registration = pluk_adapters::ToolRegistration { name: self.prefix(&registration.name), ..registration };
        self.inner.register_tool(registration, handler);
    }

    fn register_prompt(
        &mut self,
        name: &str,
        description: &str,
        args_schema: Option<serde_json::Map<String, serde_json::Value>>,
        handler: pluk_adapters::PromptHandler,
    ) {
        self.inner.register_prompt(&self.prefix(name), description, args_schema, handler);
    }

    fn register_resource(
        &mut self,
        name: &str,
        uri: &str,
        mime_type: &str,
        description: Option<&str>,
        handler: pluk_adapters::ResourceHandler,
    ) {
        let uri = namespace_uri(&self.ns, uri);
        self.inner.register_resource(&self.prefix(name), &uri, mime_type, description, handler);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::Arc;
    use pluk_adapters::{PromptHandler, ResourceHandler, ToolHandler};

    #[derive(Default)]
    struct RecordingHost {
        calls: Vec<String>,
    }

    impl ToolHost for RecordingHost {
        fn register_tool(&mut self, registration: pluk_adapters::ToolRegistration, _handler: ToolHandler) {
            self.calls.push(format!("tool:{}", registration.name));
        }

        fn register_prompt(&mut self, name: &str, _description: &str, _args_schema: Option<serde_json::Map<String, serde_json::Value>>, _handler: PromptHandler) {
            self.calls.push(format!("prompt:{name}"));
        }

        fn register_resource(&mut self, name: &str, uri: &str, _mime_type: &str, _description: Option<&str>, _handler: ResourceHandler) {
            self.calls.push(format!("res:{name}:{uri}"));
        }
    }

    fn noop_tool() -> ToolHandler {
        Arc::new(|_| Box::pin(async { unreachable!() }))
    }

    #[test]
    fn slug_makes_a_tool_safe_prefix() {
        assert_eq!(slug("Metrics DB"), "metrics_db");
        assert_eq!(slug("DB — Production!"), "db_production");
        assert_eq!(slug(""), "member");
        assert_eq!(slug("__--__"), "member");
    }

    #[test]
    fn namespaced_host_prefixes_names_and_uris() {
        let mut fake = RecordingHost::default();
        {
            let mut host = NamespacedHost::new(&mut fake, "metrics_db");
            host.register_tool(pluk_adapters::ToolRegistration::no_args("query", "Q"), noop_tool());
            host.register_prompt("summarize_schema", "S", None, Arc::new(|_| Box::pin(async { unreachable!() })));
            host.register_resource("schema", "schema://full", "text/plain", None, Arc::new(|| Box::pin(async { unreachable!() })));
        }
        assert_eq!(
            fake.calls,
            [
                "tool:metrics_db__query".to_string(),
                "prompt:metrics_db__summarize_schema".to_string(),
                "res:metrics_db__schema:schema://metrics_db/full".to_string(),
            ]
        );
    }

    #[test]
    fn uris_without_a_scheme_get_a_plus_namespace() {
        assert_eq!(namespace_uri("a", "table/users"), "a+table/users");
    }
}
